### DISCUSSION TITLE
What response does WebSend expect?

### DISCUSSION THREAD

**Discussion by afgcons** [arendst/Tasmota#20319] [Q&A] [answered]
**Device:  ZB-GW03-V1.2 ZigBee Gateway, Tasmota: 12.3.1(tasmota), Pre-compiled 2023-02-15T17:18:19**

### the rule:
on zbreceived#0xDC3A#Endpoint do WebSend [usbnas.loc/cgi-bin/mocha.cgi] zbsw3x%value%_0 endon

### now clicking button 3 on Moes ESW-0ZAA-EU Switch:
MQT: tele/zb1/SENSOR = {"ZbReceived":{"0x2F5B":{"Device":"0x2F5B","Name":"zb2_4btnSw","0006!FD":"00","LidlPower":0,"Endpoint":3,"LinkQuality":87}}}

### the rule fires and the cgi script receives and processes the payload:
RUL: ZBRECEIVED#0X2F5B#ENDPOINT performs "WebSend [usbnas.loc/cgi-bin/mocha.cgi] zbsw3x3_0"

### below the response (cgi script had returned a '\n' response):
16:10:40.805 MQT: stat/zb1/RESULT = {"WebSend":"Connect failed"}       <-- not true

### My Question:
I tried std "200" responses, with and without plain text, json format. The least evil ends with a WebSend failure. The worst spews  tons of messages, then reset and reboot. What response does WebSend expect?

NB: Other WebSend to Tasmota devices work well.

**Comment by sfromis** [comment]
If you get the failure *after some delay*, it may be a timeout, which could happen if your cgi script did not complete normally to close the connection. And it should indeed works if the server is following the usual http protocol, including but not limited to using status code 200.

Beware that if your server expects https instead of http, you should use WebQuery with the URL in regular format, instead of the weird format used with WebSend.

**Comment by afgcons** [comment, marked as answer]
It works now. Thanks for your help!

The web server is httpd running inside a matchbox size USB 5 Volt powered NS-320U Multi-Function USB NAS server (snake/os: Linux 2.6.16 armv4l). Why I didn't think about **curl --trace** before coming here? I don't know... but after I made the cgi script look like a Tasmota device the  ZB-GW03-V1.2 gateway was happy !

Below the curl list (see attached [curl_trace.txt](https://github.com/arendst/Tasmota/files/13788739/curl_trace.txt) notice the strange codes preceding and following the json payload....
